File: day09.py
#!/usr/bin/env pipenv run python
"""Solutions to day 9 of Advent of Code"""

import re
import itertools
import collections

from get_input import get_input, line_parser

class Marble:
    __slots__ = ['value', 'head', 'tail']

    def __init__(self, value, tail=None, head=None):
        self.value = value
        self.tail = tail or self
        self.head = head or self

    def insert(self, marble):
        other = Marble(marble, tail=self, head=self.head)
        self.head, other.head.tail = other, other
        return other

    def remove(self):
        self.head.tail, self.tail.head = self.tail, self.head
        return self.value, self.head

    def move(self, n=1):
        result = self
        for _ in range(abs(n)):
            result = result.head if n >= 0 else result.tail
        return result

    def __repr__(self):
        return f"Marble({self.value}, tail=Marble({self.tail.value}), "\
               f"head=Marble({self.head.value}))"

def part1(game):
    """Solution to part 1"""
    root = Marble(0)
    scores = {e: 0 for e in range(1, game[0].players+1)}
    for elf, marble in zip(itertools.cycle(
            range(1, game[0].players+1)),
            range(1, game[0].marbles+1)):
        if marble % 23 == 0:
            val, root =  root.move(-7).remove()
            scores[elf] += marble + val
        else:
            root = root.move().insert(marble)
    return max(scores.values())

def part2(game):
    """Solution to part 2"""
    return part1([Record(game[0].players, game[0].marbles * 100)])

Record = collections.namedtuple('Record', 'players, marbles')

def parse(line):
    # \d+ players; last marble is worth \d+ points
    m = re.match(r"(\d+) players; last marble is worth (\d+) points", line)
    return Record(*[int(g) for g in m.groups()])

if __name__ == '__main__':
    assert part1([Record(9, 25)]) == 32
    assert part1([Record(13, 7999)]) == 146373
    assert part1([Record(17, 1104)]) == 2764
    assert part1([Record(21, 6111)]) == 54718
    assert part1([Record(30, 5807)]) == 37305
    LINES = line_parser(get_input(day=9, year=2018), parse=parse)
    print("Part 1: {}".format(part1(LINES)))
    print("Part 2: {}".format(part2(LINES)))

File: day09.rs
extern crate common;

use common::get_input;
use regex::Regex;

use std::collections::VecDeque;

struct Input {
    n_players: usize,
    n_marbles: usize,
}

trait BoardExt<T> {
    fn rotate(&mut self, steps: i32);
}

impl <T> BoardExt<T> for VecDeque<T> {
    fn rotate(&mut self, steps: i32) {
        for _ in 0..steps.abs() {
            if steps > 0 {
                let moved = self.pop_front().unwrap();
                self.push_back(moved);
            } else {
                let moved = self.pop_back().unwrap();
                self.push_front(moved);
            }
        }
    }
}


fn part1(game: &Input) -> usize {
    let mut board: VecDeque<usize> = VecDeque::with_capacity(game.n_marbles);
    board.push_front(0);
    let mut elves = vec![0; game.n_players];
    for (marble, elf) in (1..game.n_marbles).zip((0..game.n_players).cycle()) {
        if marble % 23 == 0 {
            board.rotate(-7);
            let value = board.pop_front().unwrap();
            elves[elf] += value + marble;
        } else {
            board.rotate(2);
            board.push_front(marble);
        }
    }
    *elves.iter().max().unwrap()
}

fn part2(game: &Input) -> usize {
    part1(&Input {
        n_marbles: game.n_marbles * 100,
        n_players: game.n_players,
    })
}

fn parse(input: String) -> Input {
    let regex = Regex::new(r"(\d+) players; last marble is worth (\d+) points")
        .unwrap();
    let groups = regex.captures(input.trim()).unwrap();
    Input {
        n_players: groups[1].parse().unwrap(), 
        n_marbles: groups[2].parse().unwrap()
    }
}

fn main() {
    let input = parse(get_input(09, 2018));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
